src/err.rs

```diff
@@ -0,0 +1,192 @@
+
+    #[snafu(display("Offset {}: Token `{}` is unimplemented", offset, name))]
+    UnimplementedToken { name: String, offset: u64 },
+
+    #[snafu(display(
+        "Offset {}: Failed to decode UTF-16 string, caused by: {}",
+        offset,
+        source
+    ))]
+    FailedToDecodeUTF16String { source: std::io::Error, offset: u64 },
+
+    #[snafu(display(
+        "Offset {}: Failed to decode UTF-8 string, caused by: {}",
+        offset,
+        source
+    ))]
+    FailedToDecodeUTF8String {
+        source: std::string::FromUtf8Error,
+        offset: u64,
+    },
+
+    #[snafu(display("Offset {}: Failed to decode GUID, caused by: {}", offset, source))]
+    FailedToReadGUID { source: std::io::Error, offset: u64 },
+
+    #[snafu(display("Offset {}: Failed to decode NTSID, caused by: {}", offset, source))]
+    FailedToReadNTSID { source: std::io::Error, offset: u64 },
+
+    #[snafu(display("Failed to create record model, reason: {}", message))]
+    FailedToCreateRecordModel { message: String },
+
+    /// Errors related to Serialization
+    // Since `quick-xml` maintains the stack for us, structural errors with the XML
+    // Will be included in this generic error alongside IO errors.
+    #[snafu(display("Writing to XML failed with: {}", source))]
+    XmlOutputError { source: QuickXmlError },
+
+    #[snafu(display("Building a JSON document failed with message: {}", message,))]
+    JsonStructureError { message: String },
+
+    #[snafu(display("`serde_json` failed with error: {}", source))]
+    JsonError { source: serde_json::error::Error },
+
+    #[snafu(display("Record data contains invalid UTF-8: {}", source))]
+    RecordContainsInvalidUTF8 { source: std::string::FromUtf8Error },
+
+    /// Misc Errors
+    #[snafu(display("Unimplemented: {}", name))]
+    Unimplemented { name: String },
+    #[snafu(display("An unexpected error has occurred: {}", detail))]
+    Any { detail: String },
+}
+
+/// Generic error handler for quick prototyping, inspired by failure's `format_err!` macro.
+#[macro_export]
+macro_rules! format_err {
+   ($($arg:tt)*) => { $crate::err::Any { detail: format!($($arg)*) }.fail() }
+}
+
+/// Errors on unimplemented functions instead on panicking.
+#[macro_export]
+macro_rules! unimplemented_fn {
+   ($($arg:tt)*) => { $crate::err::Unimplemented { name: format!($($arg)*) }.fail() }
+}
+
+/// Adapter for `quick-xml` error type, which is implemented internally in `failure`,
+/// and provides no easy way of producing std compatible `Error`
+#[derive(Debug)]
+pub struct QuickXmlError {
+    message: String,
+}
+
+impl std::error::Error for QuickXmlError {
+    fn description(&self) -> &str {
+        &self.message
+    }
+}
+
+impl std::fmt::Display for QuickXmlError {
+    fn fmt(&self, f: &mut Formatter) -> std::result::Result<(), std::fmt::Error> {
+        f.write_str(&self.message)?;
+        Ok(())
+    }
+}
+
+impl From<quick_xml::Error> for QuickXmlError {
```

Comment:
It's not possible, because of the way the macro works it's not actually `Error` but the context-selector `IOErr` (which is only expanded at compile time). And it's only possible to create with the context function.

In the `quick-xml` case I still need to `.map_err` all the quick xml errors to `QuickXmlError`, and then call `.context`.